### ISSUE TITLE
Question about arguments that get passed to inner objects

### ISSUE THREAD

**Issue by somospocos** [tobi-wan-kenobi/bumblebee-status#548]
### Question

This is a question about an internal code design choice that I don't understand yet.

**bumblebee-status** has a **-t**/**--theme** command line argument. Technically this means that theme info is part of configuration. Indeed, [here](https://github.com/tobi-wan-kenobi/bumblebee-status/blob/75cf28a1ea4e481dcc8c8a8c8e7edae4e7559380/bumblebee-status#L41) I can see that **theme** is a part of **bumblebee.config.Config** data structure.

But it is computed only to be passed as argument to [bumblebee.output.I3BarOutput](https://github.com/tobi-wan-kenobi/bumblebee-status/blob/75cf28a1ea4e481dcc8c8a8c8e7edae4e7559380/bumblebee-status#L42) and [bumblebee.engine.Engine](https://github.com/tobi-wan-kenobi/bumblebee-status/blob/75cf28a1ea4e481dcc8c8a8c8e7edae4e7559380/bumblebee-status#L47). It is not used anywhere else.

This looked like a slight redundancy problem when I did #545, but now when I got out of that context, in the bigger picture this looks like a bigger problem, since this has propagated down to **bumblebee.output.WidgetDrawer**.

@tobi-wan-kenobi could you please give some details on this? Is this some deliberate design choice for a reason that so far escapes me, or is it some legacy redundancy that could use some additional refactoring?

**Comment by tobi-wan-kenobi** [comment]
The `Config` class does two things (and yeah, that's bad already - SRP). It collects commandline arguments and "merges" them with information retrieved from an optional configuration file (hence the name).

The only reason the Config class contains the *name* of the theme is the fact that this is where all CLI parameters live. The actual instantiation of the theme is done in the main( ) method of bumblebee-status anyhow.

I'm not sure I get the redundancy part, at the moment...

**Comment by somospocos** [comment]
If there's a pattern like this:

    theme = bumblebee.theme.Theme(config.theme(), config.iconset())
    InnerClass(config, theme)

I am wondering why not passing just the config instance as argument and compute the theme inside it like this:

    InnerClass(config)
    ...
    class InnerClass:
        def __init__(self, config):
            self._config = config  # if it's needed for more than finding the theme
            self._theme = bumblebee.theme.Theme(config.theme(), config.iconset())

**Comment by tobi-wan-kenobi** [comment]
Ah, I see.

That was intentional - with the idea that the theme should depend on the name of the theme, with an overridable iconset, but nothing more, ie. it should not have access to the full config.

Having said that, I have no strong feelings about that either way :) If I had to redesign this (wait - I *am* :P ), I'd probably think about just passing in the theme name and then making it possible to override certain aspects from the outside, but again: Just an idea, nothing fixed.

So, to answer your question: Yes, this was done intentionally, but with little attachment to the solution.

**Comment by somospocos** [comment]
> If I had to redesign this (wait - I am :P ), I'd probably think about **just passing in the theme name**

The part with bold emphasis is what gets me confused. Going back to my hypothetical example above, currently the Theme instance is passed as argument to InnerClass, not a theme name string.

But the Theme instance is created with a theme name string and an icon set name string as arguments.

Howerver, both the theme name string and the icon set name string already belong to the Config instance and can be accessed via Config.theme() and Config.iconset() whenever it's needed.

As my example further shows, why not just passing the Config instance to InnerClass and create the Theme instance inside it, rather than creating it outside and passing it to InnerClass together with Config instance?

Just in case if we are thinking about different things, perhaps I should just have a go at the code in another refactoring branch, and if I manage to do the above outlined conversion without breaking the code, we could continue the talk using that as reference - because "talk without code is cheap" (tm)?

I'm willing to try this, so you won't need to shift focus too much from the development branch, and we can continue this topic once I succeed and I get some code to show (or once I'll admit I'm stuck and I failed, so you could intervene).

**Comment by tobi-wan-kenobi** [comment]
Yes, having a concrete code example of what you are thinking about would be helpful.

My main concern - if I get your point correctly - was to first create all base objects (input, output, theme, etc) and then "arrange" them. I was a bit concerned about ending up with a tangle of interrelated objects with no clear ownership and logic which component creates which.

But looking at it from a distance, having the theme be a direct member of the output class sounds logical.

**Comment by somospocos** [comment]
I am still approaching this from a strictly refactoring oriented point of view rather than a design one.

In pseudo-code, to me it looks like this:

- there are two pieces of data - the config instance and the theme instance that get passed as arguments to an inner class instance
- but the data needed to create the theme instance is a subset of config instance
- to me this makes passing the theme instance redundant, one can only pass the config instance to the inner class and build the theme instance there

> My main concern was to first create all base objects (input, output, theme, etc) and then "arrange" them.

See what I said above about the theme instance: it is created in **bumblebee-status.main()**, but only for the sake of being passed to other objects as argument: **bumblebee.output.I3BarOutput** and **bumblebee.engine.Engine**. All the data needed to create the theme instance is contained in the config instance. And **both** those objects accept config as argument anyway, that's why I see no point to explicitly instantiate the theme. It would be justified if **bumblebee-status** would use it directly somewhere, ATM it doesn't.

But I'll have one more code review from bumblebee-status down to bumblebee.output.WidgetDrawer and then try to make some changes. Ideally, I might have something to show even later today, but in case I discover some unexpected troubles, I'll surely aim to get some feedback until the end of the week, when you're pressured a bit less by free time constraints.

I'll report back later in this issue.

**Comment by somospocos** [comment]
So far it's not going too well. I haven't pushed anything to github, actually I haven't even commited anything locally yet, so I'll just show a diff against current master 75cf28a :

    diff --git a/bumblebee-status b/bumblebee-status
    index bb31f94..a331369 100755
    --- a/bumblebee-status
    +++ b/bumblebee-status
    @@ -4,7 +4,6 @@ import os
     import sys
     import logging
     import signal
    -import bumblebee.theme
     import bumblebee.engine
     import bumblebee.config
     import bumblebee.output
    @@ -38,8 +37,7 @@ def main():
                     filename=config.logfile()
                 )
     
    -    theme = bumblebee.theme.Theme(config.theme(), config.iconset())
    -    output = bumblebee.output.I3BarOutput(theme=theme, config=config)
    +    output = bumblebee.output.I3BarOutput(config)
         inp = bumblebee.input.I3BarInput()
         engine = None
     
    @@ -48,7 +46,6 @@ def main():
                 config=config,
                 output=output,
                 inp=inp,
    -            theme=theme,
             )
             signal.signal(10,sig_USR1_handler)
             engine.run()
    diff --git a/bumblebee/engine.py b/bumblebee/engine.py
    index 1159884..31aa4f6 100644
    --- a/bumblebee/engine.py
    +++ b/bumblebee/engine.py
    @@ -8,6 +8,7 @@ import logging
     import importlib
     import bumblebee.error
     import bumblebee.modules
    +import bumblebee.theme
     
     log = logging.getLogger(__name__)
     
    @@ -142,16 +143,16 @@ class Engine(object):
         This class connects input/output, instantiates all
         required modules and drives the "event loop"
         """
    -    def __init__(self, config, output=None, inp=None, theme=None):
    -        self._output = output
    +    def __init__(self, config, output=None, inp=None):
             self._config = config
    +        self._output = output
             self._running = True
             self._modules = []
             self.input = inp
    +        self._theme = bumblebee.theme.Theme(config.theme(), config.iconset())
             self._aliases = self._aliases()
             self.load_modules(config.modules())
             self._current_module = None
    -        self._theme = theme
     
             if bumblebee.util.asbool(config.get("engine.workspacewheel", "true")):
                 if bumblebee.util.asbool(config.get("engine.workspacewrap", "true")):
    diff --git a/bumblebee/output.py b/bumblebee/output.py
    index b2df265..ae1bf9e 100644
    --- a/bumblebee/output.py
    +++ b/bumblebee/output.py
    @@ -9,6 +9,7 @@ import logging
     
     import bumblebee.store
     import bumblebee.util
    +import bumblebee.theme
     
     MAX_PERCENTS = 100.
     CHARS = 8
    @@ -327,12 +328,12 @@ class WidgetDrawer(object):
             Wrapper for I3BarOutput.draw(),
             because that function is getting too big
         """
    -    def __init__(self, theme, config=None):
    +    def __init__(self, config):
             """
                 Keep the same signature as I3BarOutput.__init__()
             """
    -        self._theme = theme
             self._config = config
    +        self._theme = bumblebee.theme.Theme(config.theme(), config.iconset())
             self._widgets = []
             self._markup = None
             self._full_text = None
    @@ -435,11 +436,11 @@ class WidgetDrawer(object):
     
     class I3BarOutput(object):
         """Manage output according to the i3bar protocol"""
    -    def __init__(self, theme, config=None):
    -        self._theme = theme
    +    def __init__(self, config):
    +        self._config = config
    +        self._theme = bumblebee.theme.Theme(config.theme(), config.iconset())
             self._widgets = []
             self._started = False
    -        self._config = config
     
         def started(self):
             return self._started
    @@ -461,7 +462,7 @@ class I3BarOutput(object):
                 blocks of JSON text to self._widgets: one for separator, if the
                 theme contains a separator and one for the widget itself
             """
    -        widget_drawer = WidgetDrawer(self._theme, self._config)
    +        widget_drawer = WidgetDrawer(self._config)
             self._widgets.extend(widget_drawer.draw(widget, module, engine))
     
         def begin(self):
This doesn't produce the results I've expected:

1. It breaks some tests - expectable, since certain callable signatures changed, I removed the theme argument in a few places.

2. Running this new code with the newest nord-powerline theme produces different results: there's no background in the i3bar and the cycle effect is gone.

Something is wrong with the change above, but so far I'm failing to figure out what exactly.

**Comment by somospocos** [comment]
The change above is actually still big, it can be split into smaller parts. Looks like even the smallest, deepest change can reproduce the same issue:

    diff --git a/bumblebee/output.py b/bumblebee/output.py
    index b2df265..eb1aab9 100644
    --- a/bumblebee/output.py
    +++ b/bumblebee/output.py
    @@ -9,6 +9,7 @@ import logging
     
     import bumblebee.store
     import bumblebee.util
    +import bumblebee.theme
     
     MAX_PERCENTS = 100.
     CHARS = 8
    @@ -327,12 +328,12 @@ class WidgetDrawer(object):
             Wrapper for I3BarOutput.draw(),
             because that function is getting too big
         """
    -    def __init__(self, theme, config=None):
    +    def __init__(self, config=None):
             """
                 Keep the same signature as I3BarOutput.__init__()
             """
    -        self._theme = theme
             self._config = config
    +        self._theme = bumblebee.theme.Theme(config.theme(), config.iconset())
             self._widgets = []
             self._markup = None
             self._full_text = None
    @@ -461,7 +462,7 @@ class I3BarOutput(object):
                 blocks of JSON text to self._widgets: one for separator, if the
                 theme contains a separator and one for the widget itself
             """
    -        widget_drawer = WidgetDrawer(self._theme, self._config)
    +        widget_drawer = WidgetDrawer(self._config)
             self._widgets.extend(widget_drawer.draw(widget, module, engine))
     
         def begin(self):

Here I only change the signature of WidgetDrawer by only passing the config instance and instantiating the theme object inside it, yet it's not the same - background and cycler colors are still broken in i3bar.

**Comment by tobi-wan-kenobi** [comment]
OK, that is decidedly weird. The class should benself contained :-/ You're certainly digging around in the darkest corners of the project ;-)

**Comment by somospocos** [comment]
So far I am not yet despaired enough to resort to low level step-by-step debugging. For now I have a totally fact-lacking guess why this might happen. Since the cycle colors are broken by this change, I searched where and how are they used. There is a **bumblebee.theme.Theme.reset()** call in **bumblebee.output.I3BarOutput.begin()**. While the docstring of **Theme** class advertises it as "a collection of icons and colors" - I understand it as "static and immutable object wrapper of JSON-formated dictionary data structure stored in the theme file", what's happening in **Theme.reset()** seems to prove the docstring misleading, because the Theme object also seems to track state that is used during draw iterations. I'm not sure this is good, I'd rather have the drawers (**WidgetDrawer.draw()** and **I3BarOutput.draw()**) track this state of current cycle color outside the **Theme** object. This reminds me of the classic advertised advantage of functional programming that is free from this kind of bugs, since it works with immutable data. But you should take this last one with a really big grain of salt, because I'm not a FP layman, I'm worse - I'm afraid that the count of my failed attempts to learn the basics of Haskell approaches a dozen :D

**Comment by tobi-wan-kenobi** [comment]
Ah! Yeah, I overlooked that you have independent theme instances for output and the widget drawer.

While I see the appeal of functional programming (having failed to grasp Lisp multiple times :-P), I do subscribe to the OOP state+behaviour.

The docstring of Theme is misleading, yes, but the tracking of the cycle, to my mind, belongs in the theme. Something outside should not care how the background/foreground color was derived, I think (but not a strong opinion) I'd be hard pressed to say where to draw the line then - oberriding icons, colors for modules/stati...

The fact that I'm using the reset() is the issue, in my mind, because it adds unnecessary responsibility - and required knowhow - to the caller.

In a hypothetical rewrite ;-) I'd like to have an event system that for instance emits "line start" "line end" events and the theme objects subscribe to those. That would make your scenario work - independent theme instances behave identical - and remove the reset() kludge. What is your take on that?

Cheers!

**Comment by somospocos** [comment]
> I overlooked that you have independent theme instances for output and the widget drawer.

All I wanted was to cut a corner and get rid of explicitly passing the theme argument around. But I have to admit that after diving a bit deeper into the code it looks like I based my plan on a fact that ended being false: initially I assumed that a Theme object is just a container for JSON data from a theme file (whose name is contained in Config anyway), but as the wall of text above proves, it's more than that, because it also carries state info. It looks like I need to accept the responsibility about my initial mistake and give up on the whole idea. Passing the theme object around proved to be a necessity, not a redundancy, like I initially thought.

> In a hypothetical rewrite ;-) I'd like to have an event system that for instance emits "line start" "line end" events and the theme objects subscribe to those. That would make your scenario work - independent theme instances behave identical - and remove the reset() kludge. What is your take on that?

My experience with event driven code is limited to a couple of trivial things written using "hello world"-level Tkinter where I attached simple event handlers to button click events. So I'm quite confident that I am neither able to implement a trivial event system from scratch (because I'm really not that skilled to think in abstract design patterns like those from the Gang of Four book), nor hook some existing event library into the bumblebee-status code - (this would add an external dependency anyway?) I already feel embarrassed that going a bit deeper into the bumblebee core proved to be orders of magnitude harder (my fault entirely, it just looks like I took a bite more than I could swallow, but it was worth it) than just writing a new module from scratch by using the exposed high level API.

TLDR: I am morally ready to give up on this whole additional refactoring idea and try making my tiny icons bigger on top of existing master branch.

**Comment by tobi-wan-kenobi** [comment]
My 2c: That would be a real pity, the work you did so far is really great and improving bumblebee. I'd really like to encourage you to push on, if you feel like it. You're doing good.

The fact that you encounter these issues says more about the (lack of) quality of this particular code than your coding skills.

I took care to make writing modules as painless as possible, I did not exercise the same care in the core.

The event stuff sounds more fancy than it is - just a register() and a notify() and a dict of callbacks in the background. If you look at input.py in the development branch, it's all of 50 lines of code (and should really be renamed event.py and generalized a bit)

Thanks!

**Comment by somospocos** [comment]
> My 2c: That would be a real pity, the work you did so far is really great and improving bumblebee. I'd really like to encourage you to push on, if you feel like it. You're doing good.

I really see no drama here, I believe I honestly invested enough time into trying to understand the code and making the changes I want to the point where I get confident I can't do what I want with my current level of knowledge. I see this as something normal, it's not some global surrender. I'm just doing a mental "git stash" if you want, so I can shift the focus on the more stringent practical need that haunts me (those pesky tiny icons!). This issue can stay open if you're planning to use it as reference for some stuff that you do in the development branch or if you foresee that this actual change to an event-driven system, which you consider better (seriously, I lack technical competence to have an opinion on this) is going to happen in the master branch at some point, before the development branch will get merged.

> The fact that you encounter these issues says more about the (lack of) quality of this particular code than your coding skills.
> I took care to make writing modules as painless as possible, I did not exercise the same care in the core.

I fairly think the code is fine, the main argument backing it up is that bumblebee-status works.

I just think this particular optimization is premature - not in the classical algorithm-centric meaning, but in context of it not being suitable to be done by __me__ and done __now__.

> The event stuff sounds more fancy than it is - just a register() and a notify() and a dict of callbacks in the background. If you look at input.py in the development branch, it's all of 50 lines of code (and should really be renamed event.py and generalized a bit)

I had a look at that code and then took some time to think about an analogy about the difference between me and you looking at that code. If you give me a pair of chopsticks and expect me to eat rice with them - I'd have to choose between eating with bare hands or asking for a spoon instead, because I've never used chopsticks. While you might not be able to catch flies with chopsticks like Musashi Miyamoto (that would be some Martin Fowler or Alex Martelli skill level), you're perfectly able to just eat rice with chopsticks - after all, you wrote that code, so you're able to read it as well! - while at this moment I'm still the guy who needs the spoon :-)

**Comment by tobi-wan-kenobi** [comment]
Sounds totally fine for me - and thanks for the PR!

I had a good laugh over your analogy :-P Though I do software development since 20-odd years now, I often feel like the guy who eats sushi with chopsticks, until at some point in time, a waiter politely approached "would like a fork, sir?"

Anyhow, I wish you a lot of fun tinkering with the bumblebee-status code, going forward, and feel free to approach me with any question you might have.

Cheers!

**Comment by tobi-wan-kenobi** [comment]
@somospocos Talking about the "event" system earlier: If you are interested, you might want to have another look at the development branch. There's a core.event now, and both the cycling of the theme and actually the whole output is now governed by this light-weight "event-driven" approach.

Cheers,

.t

**Comment by somospocos** [comment]
@tobi-wan-kenobi Thanks. I had a look and here's my short feedback.

Part 1 (not bumblebee related): after skimming through the code, I asked myself two questions:

1. Hypothetically, would I be able to implement an event-driven prototype that solves some practical real problem (without looking at your code at all)?

2. Again, hypothetically, would I be able to refactor the master branch to use the event-driven model (without looking at your code at all)?

Answer to both question is "no".

I did a few searches and I found that some refer this program design approach as "Observer pattern" while in other places it is referred as "Publisher pattern". Already good, I should probably dust off my "Gang of Four" copy and have a look there. As for getting to actually grok it - slow thinkers like me have it harder. I usually try over and over with incremental breaks in-between. I don't set goals like "I have X amount of time to learn Y".

Part 2 (bumblebee related): I tried to run the development branch (it's the thing I actually started with) and it already works (great, I believe it was the author of dwm who once said in an interview something like "if your software isn't usable from day 1, your design is broken"). Once the work is finished after you reach all the goals from NOTES, even if the next gen modules will somehow explicitly need to make use of events (as you can see, at this point I don't even understand if this will stay totally transparent at module level or not), I suppose the module writing wiki page will be completed with the relevant info for module writers. Meanwhile I will try to keep an eye on development branch, albeit with slightly lower priority, because the master branch currently suits my needs as end user (I am so selfish :D).

**Comment by tobi-wan-kenobi** [comment]
@somospocos 

Reg Part 1: Heh, I like your perseverance :) TBH, especially with the Publish/Subscribe pattern, a good part of the complexity is the language of the GoF book. If you think of it in implementational terms (especially in Python), it's simple. You have a dictionary, where the **key** is the event identifier (e.g. just a string or an enum or whatever), and the **value** is a list of callbacks (in C/C++ that's function pointers, in Python a lambda or just the function).

Everytime some component invokes "trigger", this does a simple lookup in the dict, iterates over all registered callbacks and executes all of them (in type-safe languages like C++, the only real catch is that the functions need to have identical arguments, but for that, there's often a generic "Event" type).

Then, the program basically has 2 stages:
1. Setup - every component registers their callbacks
2. The event loop - something that invokes the triggers as needed.

In the case of bumblebee-status, the event loop is simply a for loop that triggers the "update" event every N seconds.

So, really, it boils down to a dictionary of lists of callbacks :)

Part 2: Thanks, I am glad to hear it, that was exactly what I was going for (plus, obsessively high test coverage from the beginning).
